Add option to UTF8 Encode All Objects

OliverVahey opened the pull request:
Added UTF8 support for importing and exporting all objects from Access

Commit:
Add UTF8 Support for all Object types on Import

Commit:
Add Bulk Conversion to UTF8 for all file types

Commit:
Add UTF8 support for all object types on export

gilbertbw commented on Modules/modExtendSaveAsText.ACM:
```diff
@@ -154,65 +154,81 @@ End Sub
+      
+      Case Else
+        If ConvFileUTF8_to_ANSI(strFilename, strFilename & ".tmp") Then
+          strFilename = strFilename & ".tmp"
```
I would give this variable a different name, maybe `strTemp` or even `str`. I found it confusing that `strFilename` is either the original file name, or the temp file path.

OliverVahey replied:
done

gilbertbw commented on Modules/modFileEncoding.ACM:
```diff
@@ -521,4 +521,15 @@ Private Function bit16ShiftR(i As Long, s As Integer) As Long
   bit16ShiftR = bit16ShiftR And 65535 '&HFFFF ' ensure we only have least significant 16 bits
+End Function
+
+Function ConvFileUTF8_to_ANSI(strFilename As String, Optional strNewFilename As String) As Boolean
```
Should this be private?

OliverVahey replied:
No, it's called in modExtendSaveAsText so it needs to be public

gilbertbw replied:
OK

gilbertbw commented on Forms/Access SVN - Options.ACF:
```diff
@@ -1,4 +1,4 @@
-Version =20
```
You have reencoded this file!

OliverVahey replied:
done

gilbertbw commented on Forms/Access SVN - Options.ACF:
```diff
@@ -2419,14 +2419,14 @@ Begin Form
+                    Height =240
                     Name ="lblUTF8"
-                    Caption ="Convert Encoding to UTF-8"
+                    Caption ="Convert to UTF-8 (Old)"
```
Maybe "Deprecated" rather than "Old"?

OliverVahey replied:
done

gilbertbw commented on Modules/modExtendSaveAsText.ACM:
```diff
@@ -133,21 +139,50 @@ End_Processing:
+        End If
       End If
     Case asObjectType.acModule
       Application.SaveAsText objType, objName, strFilename ' no conversion needed - modules are output as ANSI
```
this comment is no longer correct!

OliverVahey replied:
done

gilbertbw commented on Modules/modExtendSaveAsText.ACM:
```diff
@@ -133,21 +139,50 @@ End_Processing:
+        If Not ConvFile_to_UTF8(strFilename) Then
+          MsgBox "Unable to convert '" & objName & "' to UTF-8 encoding. Beware when committing."
+        End If
       End If
     Case asObjectType.acModule
       Application.SaveAsText objType, objName, strFilename ' no conversion needed - modules are output as ANSI
+      If (enmEncode = asEncode_All_to_UTF8) Then
+        If Not ConvFile_to_UTF8(strFilename) Then
+          MsgBox "Unable to convert '" & objName & "' to UTF-8 encoding. Beware when committing."
+        End If
+      End If
     Case asObjectType.acTable, asObjectType.acIMEXSpecs
       SaveTableAsText objName, strFilename
+      If (enmEncode = asEncode_All_to_UTF8) Then
+        If Not ConvFile_to_UTF8(strFilename) Then
+          MsgBox "Unable to convert '" & objName & "' to UTF-8 encoding. Beware when committing."
+        End If
+      End If
     Case asObjectType.acToolbar
       SaveCommandBarAsText objName, strFilename
+      If (enmEncode = asEncode_All_to_UTF8) Then
+        If Not ConvFile_to_UTF8(strFilename) Then
+          MsgBox "Unable to convert '" & objName & "' to UTF-8 encoding. Beware when committing."
+        End If
+      End If
     Case asObjectType.acExtra
       SaveExtraAsText strFilename
+      If (enmEncode = asEncode_All_to_UTF8) Then
+        If Not ConvFile_to_UTF8(strFilename) Then
+          MsgBox "Unable to convert '" & objName & "' to UTF-8 encoding. Beware when committing."
+        End If
+      End If
     Case asObjectType.acReference
       SaveReferencesAsText strFilename
+      If (enmEncode = asEncode_All_to_UTF8) Then
+        If Not ConvFile_to_UTF8(strFilename) Then
+          MsgBox "Unable to convert '" & objName & "' to UTF-8 encoding. Beware when committing."
+        End If
+      End If
   End Select
```
Do you think these should be refactored?

gilbertbw replied:
Done

gilbertbw commented on Forms/Access SVN - Options.ACF:
```diff
@@ -2740,6 +2793,50 @@ Private Sub cmdUTF8Encoding_Click()
+  SetConfigValue "UTF-8 Encode All", True, Me.txtOutputLocation.Value
+    
+  Me.cmdRun.SetFocus
+  Me.cmdUTF8FullEncode.Enabled = False
```
The 'convert encoding to UTF-8' button should also be disabled, as doing this implies reencoding ucs2 files. The 2 buttons should also be disabled when you open the form if `ConfigValue "UTF-8 Encode All"` is True.

OliverVahey replied:
done

gilbertbw replied:
You still need to do the second half of this

Commit:
Improved code clarity in LoadFromText function

Commit:
Updated Label on the old UTF8 conversion button

Commit:
UFT8 Conversions now disable conversion buttons

Commit:
Remove comment that is no longer relevant

Commit:
Refactored SaveAsText

Refactored SaveAsText to reduce code repetition and increase clarity

Commit:
Fix UTF8 buttons Enabled on startup

Fix UTF8 Conversion buttons being incorrectly enabled on opening addin when a conversion had already been run

OliverVahey commented:
This work has been merged in, just via cherry-picking instead of PR
